src/main/java/org/myrobotlab/service/Py4j.java

```diff
@@ -1,148 +1,352 @@
     super(n, id);
   }
 
   /**
-   * start the gateway service listening on port
+   * Add a new script to Py4j default location will be in
+   * data/Py4j/{serviceName}
+   * 
+   * @param scriptName
+   *          - name of the script
+   * @param code
+   *          - code block
+   * @throws IOException
    */
-  public void start() {
-    if (gateway == null) {
-      gateway = new GatewayServer(this);
-      gateway.addListener(this);
-      gateway.start();
-      info("server started listening on %s:%d", gateway.getAddress(), gateway.getListeningPort());
-      handler = (Invoker) gateway.getPythonServerEntryPoint(new Class[] { Invoker.class });
-    } else {
-      log.info("Py4j gateway server already started");
-    }
-  }
+  public void addScript(String scriptName, String code) throws IOException {
 
-  public void newScript() {
-    newScript("script.py");
-  }  
+    File script = new File(scriptRootDir + fs + scriptName);
 
-  public void newScript(String scriptName) {
-    if (!openedScripts.containsKey(scriptName)) {
-      openScript(scriptName, "");
+    if (script.exists()) {
+      error("script %s already exists", scriptName);
+      return;
     }
-  }
 
-  public void openScript(String scriptName, String code) {
-    activeScript = scriptName;
-    openedScripts.put(scriptName, new Script(scriptName, code));
+    openedScripts.put(scriptName, new Script2(scriptName, code));
     broadcastState();
   }
 
+  /**
+   * removes script from memory of openScripts
+   * 
+   * @param scriptName
+   */
   public void closeScript(String scriptName) {
     openedScripts.remove(scriptName);
     broadcastState();
   }
-  
 
   @Override
-  public boolean preProcessHook(Message msg) {
-    // let the messages for this service
-    // get processed normally
-    if (methodSet.contains(msg.method)) {
-      return true;
+  public void connectionError(Exception e) {
+    error(e);
+  }
+
+  @Override /** TODO add a one shot addTask to call handler.setName(name) */
+  public void connectionStarted(Py4JServerConnection gatewayConnection) {
+    try {
+      log.info("connectionStarted {}", gatewayConnection.toString());
+      clients.put(getClientKey(gatewayConnection), new Py4jClient());
+
+      info("connection started");
+      invoke("getClients");
+    } catch (Exception e) {
+      error(e);
     }
+  }
 
-    // will probably need to queu this
-    if (handler != null) {
-      // TODO - determine clients are connected .. how many clients etc..
-      try {
-        handler.invoke(msg.method, msg.data);
-      } catch(Exception e) {
-        error(e);
+  @Override
+  public void connectionStopped(Py4JServerConnection gatewayConnection) {
+    info("connection stopped");
+    clients.remove(getClientKey(gatewayConnection));
+    invoke("getClients");
+  }
+
+  /**
+   * One of 3 methods supported on the MessageHandler() callbacks
+   * 
+   * @param code
+   */
+  public void exec(String code) {

```

Comment:
Either pull up to a common interface or implement Executor

Reply:
Yeah I'd like to see how deployment of Py4j and the roadmap for Jython/Python resolve first ...